pageserver/src/tenant/remote_timeline_client/download.rs

```diff
@@ -170,58 +170,64 @@ pub fn is_temp_download_file(path: &Utf8Path) -> bool {
 pub async fn list_remote_timelines(
     storage: &GenericRemoteStorage,
     tenant_id: TenantId,
-) -> anyhow::Result<HashSet<TimelineId>> {
+    cancel: CancellationToken,
+) -> anyhow::Result<(HashSet<TimelineId>, HashSet<String>)> {
     let remote_path = remote_timelines_path(&tenant_id);
 
     fail::fail_point!("storage-sync-list-remote-timelines", |_| {
         anyhow::bail!("storage-sync-list-remote-timelines");
     });
 
-    let timelines = download_retry(
-        || storage.list_prefixes(Some(&remote_path)),
-        &format!("list prefixes for {tenant_id}"),
+    let listing = download_retry_forever(
+        || storage.list(Some(&remote_path), ListingMode::WithDelimiter),
+        &format!("list timelines for {tenant_id}"),
+        cancel,
     )
     .await?;
 
-    if timelines.is_empty() {
-        anyhow::bail!("no timelines found on the remote storage")
-    }
-
     let mut timeline_ids = HashSet::new();
+    let mut other_prefixes = HashSet::new();
+
+    tracing::info!("list_remote_timelines prefixes:");
+    for p in &listing.prefixes {
+        tracing::info!("  '{p}'");
+    }
+    tracing::info!("list_remote_timelines keys:");
+    for p in &listing.keys {
+        tracing::info!("  '{p}'");
+    }
```

Comment:
This will be a storm of messages every startup. I suggest to log this at `debug!` level.
A summary log message at info! level might be ok.

Also, please use `#[instrument]` on the `list_remote_timelines` (not part of the suggestion!) fn and use structured logging.

```suggestion
    for prefix in &listing.prefixes {
        tracing::debug!(prefix, "found prefix");
    }
    for key in &listing.keys {
        tracing::debug!(key, "found key");
    }
    tracing::info!(prefix_count=listing.prefixes.len(), key_count=listings.keys.len(), "summary");
```

Reply:
Oops, this was debug code that I didn't mean to commit.  I've just removed it.